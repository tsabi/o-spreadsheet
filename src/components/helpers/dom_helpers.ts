/**
 * Return true if the event was triggered from
 * a child element.
 */
export function isChildEvent(parent: HTMLElement, ev: Event): boolean {
  return !!ev.target && parent!.contains(ev.target as Node);
}

export function gridOverlayPosition() {
  const spreadsheetElement = document.querySelector(".o-grid-overlay");
  if (spreadsheetElement) {
    const { top, left } = spreadsheetElement?.getBoundingClientRect();
    return { top, left };
  }
  throw new Error("Can't find spreadsheet position");
}

/**
 * Iterate over the dom tree starting at `el` and over all the children depth first.
 * */
export function* iterateChildren(el: Node): Generator<Node> {
  yield el;
  if (el.hasChildNodes()) {
    for (let child of el.childNodes) {
      yield* iterateChildren(child);
    }
  }
}
