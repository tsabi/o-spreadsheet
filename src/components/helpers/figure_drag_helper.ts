import { MIN_FIG_SIZE } from "../../constants";
import { Figure, PixelPosition, SheetScrollInfo } from "../../types";

export function dragFigureForMove(
  initialMousePosition: PixelPosition,
  currentMousePosition: PixelPosition,
  initialFigure: Figure,
  mainViewportPosition: PixelPosition,
  scrollInfo: SheetScrollInfo,
  frozenPaneOffset: number
): Figure {
  const initialMouseX = initialMousePosition.x;
  const mouseX = currentMousePosition.x;
  const viewportX = mainViewportPosition.x;

  const initialMouseY = initialMousePosition.y;
  const mouseY = currentMousePosition.y;
  const viewportY = mainViewportPosition.y;

  const deltaX = initialMouseX - mouseX;
  let newX = initialFigure.x - deltaX;

  // Freeze panes: always display the figure above the panes
  if (viewportX > 0) {
    const isInitialXInFrozenPane = initialFigure.x < viewportX - frozenPaneOffset;
    const isNewXInFrozenPane = newX < viewportX - frozenPaneOffset;
    const isNewXBelowFrozenPane = newX < scrollInfo.offsetX + viewportX - frozenPaneOffset;
    if (isInitialXInFrozenPane && !isNewXInFrozenPane) {
      newX += scrollInfo.offsetX;
    } else if (!isInitialXInFrozenPane && isNewXBelowFrozenPane) {
      newX -= scrollInfo.offsetX;
    }
  }
  newX = Math.max(newX, 0);

  const deltaY = initialMouseY - mouseY;
  let newY = initialFigure.y - deltaY;

  // Freeze panes: always display the figure above the panes
  if (viewportY > 0) {
    const isInitialYInFrozenPane = initialFigure.y < viewportY - frozenPaneOffset;
    const isNewYInFrozenPane = newY < viewportY - frozenPaneOffset;
    const isNewYBelowFrozenPane = newY < scrollInfo.offsetY + viewportY - frozenPaneOffset;

    if (isInitialYInFrozenPane && !isNewYInFrozenPane) {
      newY += scrollInfo.offsetY;
    } else if (!isInitialYInFrozenPane && isNewYBelowFrozenPane) {
      newY -= scrollInfo.offsetY;
    }
  }
  newY = Math.max(newY, 0);

  return { ...initialFigure, x: newX, y: newY };
}

export function dragFigureForResize(
  initialFigure: Figure,
  dirX: -1 | 0 | 1,
  dirY: -1 | 0 | 1,
  initialMousePosition: PixelPosition,
  currentMousePosition: PixelPosition
): Figure {
  const deltaX = dirX * (currentMousePosition.x - initialMousePosition.x);
  const deltaY = dirY * (currentMousePosition.y - initialMousePosition.y);

  let { width, height, x, y } = initialFigure;
  width = Math.max(initialFigure.width + deltaX, MIN_FIG_SIZE);
  height = Math.max(initialFigure.height + deltaY, MIN_FIG_SIZE);
  if (dirX < 0) {
    x = initialFigure.x - deltaX;
  }
  if (dirY < 0) {
    y = initialFigure.y - deltaY;
  }

  return { ...initialFigure, x, y, width, height };
}
