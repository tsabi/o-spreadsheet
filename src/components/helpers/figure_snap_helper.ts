import { FIGURE_BORDER_WIDTH } from "../../constants";
import { Figure, Pixel } from "../../types";

const SNAP_MARGIN: Pixel = 5;

type SnappingAxis = "top" | "bottom" | "vCenter" | "right" | "left" | "hCenter";

interface BorderPosition {
  border: SnappingAxis;
  position: Pixel;
}

export interface SnapLine {
  matchedFigs: Figure[];
  position: Pixel;
  snapOffset: number;
}

export function snapForMove(figureToSnap: Figure, otherFigures: Figure[]) {
  const snappedFigure = { ...figureToSnap };

  const verticalSnapLine = getSnapLine(snappedFigure, ["left", "right", "hCenter"], otherFigures, [
    "left",
    "right",
    "hCenter",
  ]);

  const horizontalSnapLine = getSnapLine(
    snappedFigure,
    ["top", "bottom", "vCenter"],
    otherFigures,
    ["top", "bottom", "vCenter"]
  );

  snappedFigure.x -= verticalSnapLine?.snapOffset || 0;
  snappedFigure.y -= horizontalSnapLine?.snapOffset || 0;

  return { snappedFigure, verticalSnapLine, horizontalSnapLine };
}

export function snapForResize(
  resizeDirX: -1 | 0 | 1,
  resizeDirY: -1 | 0 | 1,
  figureToSnap: Figure,
  otherFigures: Figure[]
) {
  const snappedFigure = { ...figureToSnap };

  const verticalSnapLine = getSnapLine(
    snappedFigure,
    [resizeDirX < 0 ? "left" : "right"],
    otherFigures,
    ["left", "right"]
  );
  if (verticalSnapLine) {
    if (resizeDirX > 0) {
      snappedFigure.width -= verticalSnapLine.snapOffset;
    } else if (resizeDirX < 0) {
      snappedFigure.x -= verticalSnapLine.snapOffset;
      snappedFigure.width += verticalSnapLine.snapOffset;
    }
  }

  const horizontalSnapLine = getSnapLine(
    snappedFigure,
    [resizeDirY < 0 ? "top" : "bottom"],
    otherFigures,
    ["top", "bottom"]
  );
  if (horizontalSnapLine) {
    if (resizeDirY > 0) {
      snappedFigure.height -= horizontalSnapLine.snapOffset;
    } else if (resizeDirY < 0) {
      snappedFigure.y -= horizontalSnapLine.snapOffset;
      snappedFigure.height += horizontalSnapLine.snapOffset;
    }
  }

  snappedFigure.x = Math.round(snappedFigure.x);
  snappedFigure.y = Math.round(snappedFigure.y);
  snappedFigure.height = Math.round(snappedFigure.height);
  snappedFigure.width = Math.round(snappedFigure.width);

  return { snappedFigure, verticalSnapLine, horizontalSnapLine };
}

/**
 * Get the position of borders for the given figure
 *
 * @param figure the figure
 * @param borders the list of border names to return the positions of
 */
function getFigureBordersPositions(figure: Figure, borders: SnappingAxis[]): BorderPosition[] {
  return borders.map((border) => ({ border, position: getBorderPosition(figure, border) }));
}

/**
 * Get a snap line for the given figure, if the figure can snap to any other figure
 *
 * @param snapFigure figure to get the snap line for
 * @param bordersOfSnappedFigToMatch borders of the given figure to be considered to find a snap match
 * @param otherFigures figures to match against the snapped figure to find a snap line
 * @param bordersOfOtherFigsToMatch borders of the other figures to be considered to find a snap match
 */
function getSnapLine(
  snapFigure: Figure,
  bordersOfSnappedFigToMatch: SnappingAxis[],
  otherFigures: Figure[],
  bordersOfOtherFigsToMatch: SnappingAxis[]
): SnapLine | undefined {
  const snapFigureBorders = getFigureBordersPositions(snapFigure, bordersOfSnappedFigToMatch);
  let closestSnap: SnapLine | undefined = undefined;
  for (const matchedFig of otherFigures) {
    const matchedBorders = getFigureBordersPositions(matchedFig, bordersOfOtherFigsToMatch);

    for (const snapFigureBorder of snapFigureBorders) {
      for (const matchedBorder of matchedBorders) {
        if (canSnap(snapFigureBorder.position, matchedBorder.position)) {
          const offset = snapFigureBorder.position - matchedBorder.position;

          if (closestSnap && Math.abs(offset) === Math.abs(closestSnap.snapOffset)) {
            closestSnap.matchedFigs.push(matchedFig);
          } else if (!closestSnap || Math.abs(offset) <= Math.abs(closestSnap.snapOffset)) {
            closestSnap = {
              matchedFigs: [matchedFig],
              position: matchedBorder.position,
              snapOffset: offset,
            };
          }
        }
      }
    }
  }
  return closestSnap;
}

/** Check if two borders are close enough to snap */
function canSnap(borderPosition1: Pixel, borderPosition2: Pixel) {
  return Math.abs(borderPosition1 - borderPosition2) <= SNAP_MARGIN;
}

/** Get the position of a border of a figure */
function getBorderPosition(fig: Figure, border: SnappingAxis): Pixel {
  switch (border) {
    case "top":
      return fig.y;
    case "bottom":
      return fig.y + fig.height + FIGURE_BORDER_WIDTH;
    case "vCenter":
      return fig.y + Math.ceil((fig.height + FIGURE_BORDER_WIDTH) / 2);
    case "left":
      return fig.x;
    case "right":
      return fig.x + fig.width + FIGURE_BORDER_WIDTH;
    case "hCenter":
      return fig.x + Math.ceil((fig.width + FIGURE_BORDER_WIDTH) / 2);
  }
}
