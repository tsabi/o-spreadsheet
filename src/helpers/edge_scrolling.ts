import { Viewport, Zone } from "../types";

export const MAX_DELAY = 140;
const MIN_DELAY = 20;
const ACCELERATION = 0.035;

/**
 * Decreasing exponential function used to determine the "speed" of edge-scrolling
 * as the timeout delay.
 *
 * Returns a timeout delay in milliseconds.
 */
export function scrollDelay(value: number): number {
  // decreasing exponential from MAX_DELAY to MIN_DELAY
  return MIN_DELAY + (MAX_DELAY - MIN_DELAY) * Math.exp(-ACCELERATION * (value - 1));
}

/**
 * This function will compare the modifications of selection to determine
 * a cell that is part of the new zone and not the previous one.
 */
export function findCellInNewZone(
  oldZone: Zone,
  currentZone: Zone,
  viewport: Viewport
): [number, number] {
  let col: number, row: number;
  const { left: oldLeft, right: oldRight, top: oldTop, bottom: oldBottom } = oldZone!;
  const { left, right, top, bottom } = currentZone;
  if (left != oldLeft) {
    col = left;
  } else if (right != oldRight) {
    col = right;
  } else {
    col = viewport.left;
  }
  if (top != oldTop) {
    row = top;
  } else if (bottom != oldBottom) {
    row = bottom;
  } else {
    row = viewport.top;
  }
  return [col, row];
}
