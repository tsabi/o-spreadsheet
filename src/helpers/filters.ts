import { range, UuidGenerator } from ".";
import { Cloneable, UID, Zone } from "../types/misc";

export class FilterTable implements Cloneable<FilterTable> {
  zone: Zone;
  filters: Filter[];

  constructor(zone: Zone) {
    this.filters = [];
    this.zone = zone;
    const uuid = new UuidGenerator();
    for (const i of range(zone.left, zone.right + 1)) {
      const filterZone = { ...this.zone, left: i, right: i };
      this.filters.push(new Filter(uuid.uuidv4(), filterZone));
    }
  }

  /** Get zone of the table without the headers */
  get contentZone(): Zone | undefined {
    if (this.zone.bottom === this.zone.top) {
      return undefined;
    }
    return { ...this.zone, top: this.zone.top + 1 };
  }

  getFilterId(col: number): string | undefined {
    for (let i = 0; i < this.filters.length; i++) {
      if (this.filters[i].col === col) {
        return this.filters[i].id;
      }
    }
    return undefined;
  }

  clone(): FilterTable {
    const table = new FilterTable(this.zone);
    table.filters = this.filters.map((filter) => filter.clone());
    return table;
  }
}

export class Filter implements Cloneable<Filter> {
  id: UID;
  zoneWithHeaders: Zone;

  constructor(id: UID, zone: Zone) {
    if (zone.left !== zone.right) {
      throw new Error("Can only define a filter on a single column");
    }
    this.id = id;
    this.zoneWithHeaders = zone;
  }

  get col() {
    return this.zoneWithHeaders.left;
  }

  /** Filtered zone, ie. zone of the filter without the header */
  get filteredZone(): Zone | undefined {
    const zone = this.zoneWithHeaders;
    if (zone.bottom === zone.top) {
      return undefined;
    }
    return { ...zone, top: zone.top + 1 };
  }

  clone(): Filter {
    return new Filter(this.id, this.zoneWithHeaders);
  }
}
