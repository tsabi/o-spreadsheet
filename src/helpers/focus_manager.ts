export class FocusManager {
  private focusableElement: HTMLElement | undefined = undefined;

  setFocusableElement(element: HTMLElement) {
    this.focusableElement = element;
  }

  focus() {
    this.focusableElement?.focus();
  }
}
