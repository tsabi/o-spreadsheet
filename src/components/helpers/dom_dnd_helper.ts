import { startDnd } from "./drag_and_drop";

interface DragAndDropItem {
  id: string;
  width: number;
  startingX: number;
  x: number;
}

export class DOMDndHelper {
  /**
   * The dead zone is an area in which the mousemove events are ignored.
   *
   * This is useful when swapping the dragged item with a larger item. After the swap,
   * the mouse is still hovering on the item  we just swapped with. In this case, we don't want
   * a mouse move to trigger another swap the other way around, so we create a dead zone. We will clear
   * the dead zone when the mouse leaves the swapped item.
   */
  deadZone: { start: number; end: number } | undefined;

  initialMouseX: number;
  currentMouseX: number;

  minX: number;
  maxX: number;

  constructor(
    public draggedItemId: string,
    mouseX: number,
    private items: DragAndDropItem[],
    private onChange: (newPositions: Record<string, number>) => void,
    private onCancel: () => void,
    private onDragEnd: (itemId: string, indexAtEnd: number) => void
  ) {
    this.initialMouseX = mouseX;
    this.currentMouseX = mouseX;

    this.minX = this.items[0].x;
    this.maxX = this.items[this.items.length - 1].x + this.items[this.items.length - 1].width;

    startDnd(this.onMouseMove.bind(this), this.onMouseUp.bind(this));
  }

  private onMouseMove(ev: MouseEvent) {
    if (ev.button !== 0) {
      this.onCancel();
      return;
    }
    const mouseX = ev.clientX;

    const hoveredSheetIndex = this.getHoveredItemIndex(
      mouseX,
      this.items.map((item) => item.x)
    );
    const draggedSheetIndex = this.items.findIndex((item) => item.id === this.draggedItemId);
    const draggedSheet = this.items[draggedSheetIndex];

    this.currentMouseX = mouseX;

    if (this.deadZone && mouseX >= this.deadZone.start && mouseX <= this.deadZone.end) {
      this.onChange(this.getItemsPositions());
      return;
    } else if (mouseX >= draggedSheet.x && mouseX <= draggedSheet.x + draggedSheet.width) {
      this.deadZone = undefined;
    }

    if (draggedSheetIndex === hoveredSheetIndex) {
      this.onChange(this.getItemsPositions());
      return;
    }

    const startIndex = Math.min(draggedSheetIndex, hoveredSheetIndex);
    const endIndex = Math.max(draggedSheetIndex, hoveredSheetIndex);
    const dir = Math.sign(hoveredSheetIndex - draggedSheetIndex);

    let movedWidth = 0;
    for (let i = startIndex; i <= endIndex; i++) {
      if (i === draggedSheetIndex) {
        continue;
      }
      this.items[i].x -= dir * draggedSheet.width;
      movedWidth += this.items[i].width;
    }

    draggedSheet.x += dir * movedWidth;
    this.deadZone =
      dir > 0
        ? { start: mouseX, end: draggedSheet.x }
        : { start: draggedSheet.x + draggedSheet.width, end: mouseX };
    this.items.sort((item1, item2) => item1.x - item2.x);

    this.onChange(this.getItemsPositions());
  }

  private onMouseUp(ev: MouseEvent) {
    if (ev.button !== 0) {
      this.onCancel();
      return;
    }

    const targetSheetIndex = this.items.findIndex((item) => item.id === this.draggedItemId);
    this.onDragEnd(this.draggedItemId, targetSheetIndex);
  }

  private getHoveredItemIndex(mouseX: number, xs: number[]): number {
    let hoveredItemIndex = -1;
    for (let x of xs) {
      if (x > mouseX) {
        break;
      }
      hoveredItemIndex++;
    }
    return Math.max(0, hoveredItemIndex);
  }

  private getItemsPositions() {
    const positions: Record<string, number> = {};
    for (let item of this.items) {
      if (item.id !== this.draggedItemId) {
        positions[item.id] = Math.floor(item.x - item.startingX);
        continue;
      }

      let mouseOffset = this.currentMouseX - this.initialMouseX;
      let left = mouseOffset;
      left = Math.max(this.minX - item.startingX, left);
      left = Math.min(this.maxX - item.startingX - item.width, left);

      positions[item.id] = Math.floor(left);
    }
    return positions;
  }
}
