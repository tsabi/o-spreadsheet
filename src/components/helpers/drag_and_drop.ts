import { HEADER_HEIGHT, HEADER_WIDTH, TOPBAR_HEIGHT } from "../../constants";
import { MAX_DELAY } from "../../helpers";
import { SpreadsheetChildEnv } from "../../types/env";
type EventFn = (ev: MouseEvent) => void;

export function startDnd(
  onMouseMove: EventFn,
  onMouseUp: EventFn,
  onMouseDown: EventFn = () => {}
) {
  const _onMouseUp = (ev: MouseEvent) => {
    onMouseUp(ev);

    window.removeEventListener("mousedown", onMouseDown);
    window.removeEventListener("mouseup", _onMouseUp);
    window.removeEventListener("dragstart", _onDragStart);
    window.removeEventListener("mousemove", onMouseMove);
    window.removeEventListener("wheel", onMouseMove);
  };
  function _onDragStart(ev: DragEvent) {
    ev.preventDefault();
  }
  window.addEventListener("mousedown", onMouseDown);
  window.addEventListener("mouseup", _onMouseUp);
  window.addEventListener("dragstart", _onDragStart);
  window.addEventListener("mousemove", onMouseMove);
  window.addEventListener("wheel", onMouseMove);
}

export function spreadsheetPosition() {
  const spreadsheetElement = document.querySelector(".o-spreadsheet");
  if (spreadsheetElement) {
    const { top, left } = spreadsheetElement?.getBoundingClientRect();
    return { top, left };
  }
  throw new Error("Can't find spreadsheet position");
}

/**
 * Function to be used during a mousedown event, this function allows to
 * perform actions related to the mousemove and mouseup events and adjusts the viewport
 * when the new position related to the mousemove event is outside of it.
 * Among inputs are two callback functions. First intended for actions performed during
 * the mousemove event, it receives as parameters the current position of the mousemove
 * (occurrence of the current column and the current row). Second intended for actions
 * performed during the mouseup event.
 */
export function dragAndDropBeyondTheViewport(
  env: SpreadsheetChildEnv,
  cbMouseMove: (col: number, row: number, ev: MouseEvent) => void,
  cbMouseUp: () => void,
  only: "horizontal" | "vertical" | false = false
) {
  let timeOutId: any = null;
  let currentEv: MouseEvent;
  let previousEv: MouseEvent;
  let startingEv: MouseEvent;
  let startingX: number;
  let startingY: number;
  const getters = env.model.getters;
  const sheetId = getters.getActiveSheetId();
  const position = spreadsheetPosition();

  let colIndex: number;
  let rowIndex: number;
  const onMouseDown = (ev: MouseEvent) => {
    previousEv = ev;
    startingEv = ev;
    startingX = startingEv.clientX - position.left - HEADER_WIDTH;
    startingY = startingEv.clientY - position.top - TOPBAR_HEIGHT - HEADER_HEIGHT;
  };
  const onMouseMove = (ev: MouseEvent) => {
    currentEv = ev;
    if (timeOutId) {
      return;
    }

    const { maxOffsetX, maxOffsetY } = getters.getMaximumViewportOffset(sheetId);
    const { offsetCorrectionX, offsetCorrectionY } = getters.getViewportOffsetCorrection();
    let {
      top,
      left,
      bottom,
      right,
      offsetScrollbarX: offsetX,
      offsetScrollbarY: offsetY,
    } = getters.getActiveViewport();
    const { horizontal, vertical } = getters.getPaneDivisions(sheetId);
    let canEdgeScroll = false;
    let timeoutDelay = MAX_DELAY;

    const x = currentEv.clientX - position.left - HEADER_WIDTH;
    colIndex = getters.getColIndex(x);

    if (only !== "vertical") {
      const previousX = previousEv.clientX - position.left - HEADER_WIDTH;
      const edgeScrollInfoX = getters.getEdgeScrollCol(x, previousX, startingX);
      if (edgeScrollInfoX.canEdgeScroll) {
        canEdgeScroll = true;
        timeoutDelay = Math.min(timeoutDelay, edgeScrollInfoX.delay);
        let newTarget: number;
        switch (edgeScrollInfoX.direction) {
          case "reset":
            colIndex = horizontal;
            newTarget = horizontal;
            break;
          case 1:
            colIndex = right;
            newTarget = left + 1;
            break;
          case -1:
            colIndex = left - 1;
            newTarget = left - 1;
            break;
        }
        offsetX = Math.min(
          maxOffsetX,
          getters.getColDimensions(sheetId, newTarget!).start - offsetCorrectionX
        );
      }
    }

    const y = currentEv.clientY - position.top - TOPBAR_HEIGHT - HEADER_HEIGHT;
    rowIndex = getters.getRowIndex(y);

    if (only !== "horizontal") {
      const previousY = previousEv.clientY - position.top - TOPBAR_HEIGHT - HEADER_HEIGHT;
      const edgeScrollInfoY = getters.getEdgeScrollRow(y, previousY, startingY);
      if (edgeScrollInfoY.canEdgeScroll) {
        canEdgeScroll = true;
        timeoutDelay = Math.min(timeoutDelay, edgeScrollInfoY.delay);
        let newTarget: number;
        switch (edgeScrollInfoY.direction) {
          case "reset":
            rowIndex = vertical;
            newTarget = vertical;
            break;
          case 1:
            rowIndex = bottom;
            newTarget = top + edgeScrollInfoY.direction;
            break;
          case -1:
            rowIndex = top - 1;
            newTarget = top + edgeScrollInfoY.direction;
            break;
        }
        offsetY = Math.min(
          maxOffsetY,
          env.model.getters.getRowDimensions(sheetId, newTarget!).start - offsetCorrectionY
        );
      }
    }

    cbMouseMove(colIndex, rowIndex, currentEv);
    if (canEdgeScroll) {
      env.model.dispatch("SET_VIEWPORT_OFFSET", { offsetX, offsetY });
      timeOutId = setTimeout(() => {
        timeOutId = null;
        onMouseMove(currentEv);
      }, Math.round(timeoutDelay));
    }
    previousEv = currentEv;
  };

  const onMouseUp = () => {
    clearTimeout(timeOutId);
    cbMouseUp();
  };

  startDnd(onMouseMove, onMouseUp, onMouseDown);
}
