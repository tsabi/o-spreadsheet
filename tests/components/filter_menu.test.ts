import { App } from "@odoo/owl";
import { Model, Spreadsheet } from "../../src";
import { SheetId } from "../../src/types";
import {
  createFilter,
  hideRows,
  setCellContent,
  updateFilter,
} from "../test_helpers/commands_helpers";
import { simulateClick } from "../test_helpers/dom_helper";
import { getCellContent } from "../test_helpers/getters_helpers";
import { makeTestFixture, mountSpreadsheet, nextTick } from "../test_helpers/helpers";

describe("Filter menu component", () => {
  let fixture: HTMLElement;
  let model: Model;
  let sheetId: SheetId;
  let parent: Spreadsheet;
  let app: App;

  async function openFilterMenu() {
    await simulateClick(".o-filter-icon");
  }

  function getFilterMenuValues() {
    const values: { value: string; isChecked: boolean }[] = [];
    const filterValueEls = fixture.querySelectorAll(".o-filter-menu-value");
    for (let filterValue of filterValueEls) {
      const isChecked = filterValue.querySelector(".o-filter-menu-value-checked") !== null;
      const value = filterValue.querySelector(".o-filter-menu-value-text")!.textContent!;
      values.push({ value, isChecked });
    }
    return values;
  }

  beforeEach(async () => {
    fixture = makeTestFixture();
    ({ app, parent } = await mountSpreadsheet(fixture));
    model = parent.model;
    sheetId = model.getters.getActiveSheetId();

    createFilter(model, "A1:A5");
    setCellContent(model, "A1", "header");
    setCellContent(model, "A2", "1");
    setCellContent(model, "A3", "1");
    setCellContent(model, "A4", "2");

    createFilter(model, "B1:B4");
    setCellContent(model, "B2", "B2");
    setCellContent(model, "B3", "B3");
    setCellContent(model, "B4", "B4");
    await nextTick();
  });

  afterEach(() => {
    fixture.remove();
    app.destroy();
  });

  //TODO search bar test
  //TODO skip
  test.skip("Filter menu is correctly rendered", async () => {
    await openFilterMenu();
    expect(fixture.querySelector(".o-filter-menu")).toMatchSnapshot();
  });

  test("Duplicates values are not displayed twice", async () => {
    await openFilterMenu();
    const values = getFilterMenuValues();
    expect(values.map((val) => val.value)).toEqual(["1", "2", "Blanks"]);
  });

  test("Values are checked depending on the filter state", async () => {
    updateFilter(model, "A1", ["1"]);
    await openFilterMenu();
    const values = getFilterMenuValues();
    expect(values).toEqual([
      { value: "1", isChecked: false },
      { value: "2", isChecked: true },
      { value: "Blanks", isChecked: true },
    ]);
  });

  test("Hidden values are not displayed", async () => {
    hideRows(model, [3]);
    await openFilterMenu();
    const values = getFilterMenuValues();
    expect(values.map((val) => val.value)).toEqual(["1", "Blanks"]);
  });

  test("Values hidden by another filter are not displayed", async () => {
    updateFilter(model, "B1", ["B2", "B3"]);
    await openFilterMenu();
    const values = getFilterMenuValues();
    expect(values.map((val) => val.value)).toEqual(["2", "Blanks"]);
  });

  test("Clicking on vales check and uncheck them", async () => {
    await openFilterMenu();
    expect(getFilterMenuValues()[0]).toEqual({ value: "1", isChecked: true });
    await simulateClick(".o-filter-menu-value");
    expect(getFilterMenuValues()[0]).toEqual({ value: "1", isChecked: false });
    await simulateClick(".o-filter-menu-value");
    expect(getFilterMenuValues()[0]).toEqual({ value: "1", isChecked: true });
  });

  test("Confirm button updates the filter", async () => {
    expect(model.getters.getFilter(sheetId, 0, 0)!.filteredValues).toEqual([]);
    await openFilterMenu();
    await simulateClick(".o-filter-menu-value:nth-of-type(1)");
    await simulateClick(".o-filter-menu-value:nth-of-type(2)");
    await simulateClick(".o-filter-menu-button-primary");
    expect(model.getters.getFilter(sheetId, 0, 0)!.filteredValues).toEqual(["1", "2"]);
  });

  test("Cancel button don't save the changes", async () => {
    expect(model.getters.getFilter(sheetId, 0, 0)!.filteredValues).toEqual([]);
    await openFilterMenu();
    await simulateClick(".o-filter-menu-value:nth-of-type(1)");
    await simulateClick(".o-filter-menu-value:nth-of-type(2)");
    await simulateClick(".o-filter-menu-button-cancel");
    expect(model.getters.getFilter(sheetId, 0, 0)!.filteredValues).toEqual([]);
  });

  test("Can clear and select all", async () => {
    await openFilterMenu();
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: true },
      { value: "2", isChecked: true },
      { value: "Blanks", isChecked: true },
    ]);
    await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: false },
      { value: "2", isChecked: false },
      { value: "Blanks", isChecked: false },
    ]);
    await simulateClick(".o-filter-menu-action-text:nth-of-type(1)");
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: true },
      { value: "2", isChecked: true },
      { value: "Blanks", isChecked: true },
    ]);
  });

  test("Can clear and select all", async () => {
    await openFilterMenu();
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: true },
      { value: "2", isChecked: true },
      { value: "Blanks", isChecked: true },
    ]);
    await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: false },
      { value: "2", isChecked: false },
      { value: "Blanks", isChecked: false },
    ]);
    await simulateClick(".o-filter-menu-action-text:nth-of-type(1)");
    expect(getFilterMenuValues()).toEqual([
      { value: "1", isChecked: true },
      { value: "2", isChecked: true },
      { value: "Blanks", isChecked: true },
    ]);
  });

  test("Can sort Ascending and descending", async () => {
    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(2)");
    expect(getCellContent(model, "A2")).toEqual("2");
    expect(getCellContent(model, "A3")).toEqual("1");
    expect(getCellContent(model, "A4")).toEqual("1");
    expect(getCellContent(model, "A5")).toEqual("");

    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(1)");
    expect(getCellContent(model, "A2")).toEqual("1");
    expect(getCellContent(model, "A3")).toEqual("1");
    expect(getCellContent(model, "A4")).toEqual("1");
    expect(getCellContent(model, "A5")).toEqual("");
  });
});
