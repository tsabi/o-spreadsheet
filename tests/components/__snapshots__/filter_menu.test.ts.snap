// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter menu component Filter menu is correctly rendered 1`] = `null`;
